src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts

```diff
@@ -0,0 +1,210 @@
+import type { OsmosisSupportedChainId } from './utils/types'
+
+const tradeQuoteMetadata: Map<string, { chainId: OsmosisSupportedChainId }> = new Map()
+
+export const osmosisApi: Swapper2Api = {
+  getTradeQuote: async (
+    input: GetTradeQuoteInput,
+    { sellAssetUsdRate }: { sellAssetUsdRate: string },
+  ): Promise<Result<TradeQuote2, SwapErrorRight>> => {
+    const tradeQuoteResult = await getTradeQuote(input, { sellAssetUsdRate })
+
+    return tradeQuoteResult.map(tradeQuote => {
+      const { receiveAddress, affiliateBps } = input
+      const id = uuid()
+      // TODO(gomes): getTradeQuote() is currently a hack, which represents the swap as a single trade, and needs to be revamped
+      // This effectively means that getUnsignedTx() won't be able to be implemented for all steps, but only the first for now
+      // i.e either the IBC transfer or swap-exact-amount-in
+      tradeQuoteMetadata.set(id, {
+        chainId: tradeQuote.steps[0].sellAsset.chainId as OsmosisSupportedChainId,
+      })
+      return { id, receiveAddress, affiliateBps, ...tradeQuote }
+    })
+  },
+
+  getUnsignedTx: async ({
+    from,
+    tradeQuote,
+    stepIndex,
+  }: GetUnsignedTxArgs): Promise<CosmosSignTx> => {
+    const { accountNumber, buyAsset, sellAsset, sellAmountBeforeFeesCryptoBaseUnit } =
+      tradeQuote.steps[stepIndex]
+    const { receiveAddress } = tradeQuote
+
+    const sellAssetIsOnOsmosisNetwork = sellAsset.chainId === osmosisChainId
+
+    const sellAssetDenom = symbolDenomMapping[sellAsset.symbol as keyof SymbolDenomMapping]
+    const buyAssetDenom = symbolDenomMapping[buyAsset.symbol as keyof SymbolDenomMapping]
+
+    const adapterManager = getChainAdapterManager()
+    const osmosisAdapter = adapterManager.get(osmosisChainId) as osmosis.ChainAdapter | undefined
+    const cosmosAdapter = adapterManager.get(cosmosChainId) as cosmos.ChainAdapter | undefined
+
+    const { REACT_APP_OSMOSIS_NODE_URL: osmoUrl } = getConfig()
+
+    if (!cosmosAdapter || !osmosisAdapter) throw new Error('Failed to get adapters')
+
+    let sellAddress
+
+    if (sellAssetIsOnOsmosisNetwork) {
+      sellAddress = from
+
+      if (!sellAddress)
+        throw new SwapError('failed to get osmoAddress', {
+          code: SwapErrorType.EXECUTE_TRADE_FAILED,
+        })
+    } else {
+      /** If the sell asset is not on the Osmosis network, we need to bridge the
+       * asset to the Osmosis network first in order to perform a swap on Osmosis DEX.
+       */
+      sellAddress = from
+
+      if (!sellAddress) throw new Error('Failed to get address')
+
+      const transfer = {
+        sender: sellAddress,
+        receiver: receiveAddress,
+        amount: sellAmountBeforeFeesCryptoBaseUnit,
+      }
+
+      const responseAccount = await cosmosAdapter.getAccount(sellAddress)
+      const ibcAccountNumber = parseInt(responseAccount.chainSpecific.accountNumber || '0')
```

Comment:
the quote should have the account number as typeof number (not string), which might be a less round about way of getting it

Reply:
Will tackle in the refactor follow-up, we indeed `toString()` it back in `buildPerformIbcTransferUnsignedTx`

Reply:
Tackled in https://github.com/shapeshift/web/pull/4903